### ISSUE TITLE
What GPU do you use? And how much VRAM does it have?

### ISSUE THREAD

**Issue by xinghua3632** [X-PLUG/mPLUG-Owl#35]
I just cannot start instruction tuning for stage2 on single A100 with 40GB VRAM.
TKS！

**Comment by MAGAer13** [comment]
Did you enable the bf16 and gradient checkpointing, as well as reducing the batch size?

**Comment by MAGAer13** [comment]
We are cleaning up and compatible our code to huggingface right now. So we will analysis the problem in recent few days. Once we have done, we will notice you ASAP. 😊

**Comment by xinghua3632** [comment]
I already enable bf16 and gradient checkpointing, and set batchsize=4; In your code, gradient checkpointing is only enable for CLIP， when I also enable it for LLaMA I can start the training with 26GB VRAM used. Compatible with  huggingface is a great idea. TKS!

**Comment by MAGAer13** [comment]
Thanks for reporting the potential bug to us!

**Comment by xinghua3632** [comment]
Another question, the training samples in Instruction Tuning, there are only one pair of "Human: " and "AI: ", so is it only support single-turn dialogue inputs?

**Comment by MAGAer13** [comment]
It can support the multiple turn dialogue inputs. We only calculate the loss reponsed by "AI".

**Comment by JianHu-1** [comment]
> I already enable bf16 and gradient checkpointing, and set batchsize=4; In your code, gradient checkpointing is only enable for CLIP， when I also enable it for LLaMA I can start the training with 26GB VRAM used. Compatible with huggingface is a great idea. TKS!

Hi! i encount same question! 
how to enable gradient checkpointing for LLaMA to reduce VRAM?
TKS!

**Comment by lovecambi** [comment]
> I already enable bf16 and gradient checkpointing, and set batchsize=4; In your code, gradient checkpointing is only enable for CLIP， when I also enable it for LLaMA I can start the training with 26GB VRAM used. Compatible with huggingface is a great idea. TKS!

@xinghua3632 How did you enable gradient checkpointing for LLaMA in this codebase?

**Comment by lovecambi** [comment]
> > I already enable bf16 and gradient checkpointing, and set batchsize=4; In your code, gradient checkpointing is only enable for CLIP， when I also enable it for LLaMA I can start the training with 26GB VRAM used. Compatible with huggingface is a great idea. TKS!
> 
> Hi! i encount same question! how to enable gradient checkpointing for LLaMA to reduce VRAM? TKS!

In ```train.py```, you can modify the from the [Line 111](https://github.com/X-PLUG/mPLUG-Owl/blob/main/train.py#L111). For convenience, I also add an arg ```gradient_checkpointing```.
```
    if args.use_lora:
        for param in model.parameters():
            # freeze base model's layers
            param.requires_grad = False
        peft_config = LoraConfig(
            target_modules=r'.*language_model.*\.(q_proj|v_proj)', inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.05
        )
        # lora + gradient_checkpointing
        if args.gradient_checkpointing:
            if hasattr(model, "enable_input_require_grads"):
                model.enable_input_require_grads()
            else:
                def make_inputs_require_grad(module, input, output):
                    output.requires_grad_(True)
                model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
        model = get_peft_model(model, peft_config)
        model.print_trainable_parameters()
    else:
        for name, param in model.named_parameters():
            if 'language_model' in name:
                param.requires_grad = True
            else:
                param.requires_grad = False
    if args.gradient_checkpointing:
        model.language_model.apply(
            partial(model.language_model._set_gradient_checkpointing, value=True))
```

Also, enable gradient checkpointing [here](https://github.com/X-PLUG/mPLUG-Owl/blob/main/train.py#L165).

**Comment by MAGAer13** [comment]
Hi, all! Just quick notice about our latest version of mPLUG-Owl which remove the dependency of apex! And it is more easy to use which only requires a single V100 GPU for training and fixed all of the bugs listed here!
